### PR TITLE
Implement Threads & Stages

### PR DESCRIPTION
**Pull request by Lulalaby**
Implements every aspect of stages & threads.

Threads have checks to be sure that an specific action can be executed.
Audit log is fully implemented.

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
Borderline #890 in terms of scale + @DWaffles and myself were already working on threads

**Comment by Lulalaby** [comment]
> 
> 
> Borderline #890 in terms of scale + @DWaffles and myself were already working on threads

But I'm already finished 🤷‍♀️

**Comment by Lulalaby** [comment]
The only crazy things are the first commits. I don't remember that I did it.

**Comment by Lulalaby** [comment]
[dsp.diff.txt](https://github.com/DSharpPlus/DSharpPlus/files/6887676/dsp.diff.txt)
Here is the diff

**Comment by Lulalaby** [comment]
And 
> Borderline #890 in terms of scale 

Compared to #890, this PR is quite small. In addition, only the stage & thread changes are really in here.
Sorry for the size 🤷‍♀️ but I don't think it could be smaller.

File: DSharpPlus/Clients/DiscordShardedClient.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ public DiscordShardedClient(DiscordConfiguration config)
                 this._manuallySharding = true;
 
             this.Configuration = config;
-            this.ShardClients = new ReadOnlyConcurrentDictionary<int, DiscordClient>(this._shards);
+            this.ShardClients = new ConcurrentDictionary<int, DiscordClient>(this._shards);
```
Comment (VelvetThePanda) [inline DSharpPlus/Clients/DiscordShardedClient.cs]: Same as above.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 2]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
         public Task<DiscordInvite> CreateInviteAsync(int max_age = 86400, int max_uses = 0, bool temporary = false, bool unique = false, string reason = null)
             => this.Discord.ApiClient.CreateChannelInviteAsync(this.Id, max_age, max_uses, temporary, unique, reason);
 
+
+        #region Stage
+
+        /// <summary>
+        /// Opens a stage
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:595]: Punctuation. Grammar.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 3]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// </summary>
+        /// <param name="topic">Topic of stage.</param>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:597]: Grammar (missing "the").

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 4]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <param name="privacy_level">Privacy level of stage (Defaults to <see cref="StagePrivacyLevel.GUILD_ONLY"/></param>
+        /// <param name="reason">Audit log reason</param>
+        /// <returns>Stage instance</returns>
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageChannels"/> permission.</exception>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:601]: client -> user

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 5]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task<DiscordStageInstance> OpenStageAsync(string topic, StagePrivacyLevel privacy_level = StagePrivacyLevel.GUILD_ONLY, string reason = null)
+            => await this.Discord.ApiClient.CreateStageInstanceAsync(this.Id, topic, privacy_level, reason);
+
+        /// <summary>
+        /// Modifies a stage topic
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:609]: Punctuation.

**Comment by Lulalaby** [comment]
nvm fixed

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 6]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// </summary>
+        /// <param name="topic">New topic of stage.</param>
+        /// <param name="privacy_level">New privacy level of stage.</param>
+        /// <param name="reason">Audit log reason</param>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:613]: Punctuation and grammar.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 7]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageChannels"/> permission.</exception>
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task ModifyStageAsync(Optional<string> topic, Optional<StagePrivacyLevel> privacy_level, string reason = null)
+            => await this.Discord.ApiClient.ModifyStageInstanceAsync(this.Id, topic, privacy_level, reason);
+
+        /// <summary>
+        /// Closes a stage
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:622]: Punctuation.

**Comment by Lulalaby** [comment]
[dsp-master-lulalaby.diff.txt](https://github.com/DSharpPlus/DSharpPlus/files/6887735/dsp-master-lulalaby.diff.txt)
So actual diff

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 8]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <param name="reason">Audit log reason</param>
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageChannels"/> permission.</exception>
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task CloseStageAsync(string reason = null)
+            => await this.Discord.ApiClient.DeleteStageInstanceAsync(this.Id, reason);
+
+        /// <summary>
+        /// Gets a stage
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:633]: Punctuation.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 9]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task<DiscordStageInstance> GetStageAsync()
+            => await this.Discord.ApiClient.GetStageInstanceAsync(this.Id);
+
+        #endregion
+
+        #region Threads
+
+        /// <summary>
+        /// Creates a thread. Depending on whether it is created inside an <see cref="ChannelType.News"/> or an <see cref="ChannelType.Text"/> it is either an <see cref="ChannelType.NewsThread"/> or an <see cref="ChannelType.PublicThread"/>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:648]: A thread within a thread?... That doesn't seem right.

**Comment by DWaffles** [comment]
> Borderline #890 in terms of scale + @DWaffles and myself were already working on threads

This more or less. Would have appreciated a heads up that you were working on this so duplicate effort wouldn't have happened.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 10]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the <see cref="ThreadAutoArchiveDuration"/> cannot be modified. This happens, when the guild hasn't reached a certain boost <see cref="PremiumTier"/>.</exception>
+        public async Task<DiscordThreadChannel> CreateThreadAsync(string name, ThreadAutoArchiveDuration auto_archive_duration = ThreadAutoArchiveDuration.OneHour, string reason = null) => Utilities.CheckThreadAutoArchiveDurationFeature(this.Guild, auto_archive_duration) ? await this.Discord.ApiClient.CreateThreadWithoutMessageAsync(this.Id, name, auto_archive_duration, reason) : throw new NotSupportedException($"Cannot modify ThreadAutoArchiveDuration. Guild needs boost tier {(auto_archive_duration == ThreadAutoArchiveDuration.ThreeDays ? "one" : "two")}.");
+
+        /// <summary>
+        /// Gets joined archived private threads. Can contain more threads. Watch for HasMore.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:662]: "Gets the threads the current user has joined that have been archived. HasMore will be <see langword="true"> if there are more"? I don't know what HasMore actually entails.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 11]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// </summary>
+        /// <param name="before">Get threads before this timestamp.</param>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:664]: this -> the provided

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 12]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <returns><see cref="DiscordThreadResult"/></returns>
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ReadMessageHistory"/> permission.</exception>
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task<DiscordThreadResult> GetJoinedPrivateArchivedThreadsAsync(ulong? before, int? limit)
+            => await this.Discord.ApiClient.GetJoinedPrivateArchivedThreadsAsync(this.Id, before, limit);
+
+        /// <summary>
+        /// Gets archived public threads. Can contain more threads. Watch for HasMore.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:675]: Ditto.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 13]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// </summary>
+        /// <param name="before">Get threads before this timestamp.</param>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:677]: Ditto.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs [hunk 14]
```diff
@@ -581,6 +588,117 @@ public Task<IReadOnlyList<DiscordInvite>> GetInvitesAsync()
...
+        /// <returns><see cref="DiscordThreadResult"/></returns>
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ReadMessageHistory"/> permission.</exception>
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task<DiscordThreadResult> GetPublicArchivedThreadsAsync(ulong? before, int? limit)
+            => await this.Discord.ApiClient.GetPublicArchivedThreadsAsync(this.Id, before, limit);
+
+        /// <summary>
+        /// Gets archived private threads. Can contain more threads. Watch for HasMore.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/DiscordChannel.cs:688]: Ditto.

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 15]
```diff
@@ -49,7 +49,9 @@ internal DiscordMessage()
             this._stickersLazy = new Lazy<IReadOnlyList<DiscordMessageSticker>>(() => new ReadOnlyCollection<DiscordMessageSticker>(this._stickers));
             this._jumpLink = new Lazy<Uri>(() =>
             {
-                var gid = this.Channel is DiscordDmChannel ? "@me" : this.Channel.GuildId.Value.ToString(CultureInfo.InvariantCulture);
+                var gid = this.Channel != null
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:52]: I don't see why channel would ever be null?

**Comment by Lulalaby** [comment]
> 
> 
> > Borderline #890 in terms of scale + @DWaffles and myself were already working on threads
> 
> This more or less. Would have appreciated a heads up that you were working on this so duplicate effort wouldn't have happened.

You all know that I was working on this still.
I told you that I'll make it seperate.

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs (hunk 15 above)
Comment (VelvetThePanda) [reply DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:52]: But even then, I'd prefer you do an if/else or a switch over a nested ternary. They're a pain to read.

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 16]
```diff
@@ -531,6 +551,28 @@ public Task ModifyEmbedSuppressionAsync(bool hideEmbeds)
         public Task DeleteAsync(string reason = null)
             => this.Discord.ApiClient.DeleteMessageAsync(this.ChannelId, this.Id, reason);
 
+
+        /// <summary>
+        /// Creates a thread. This thread is always a <see cref="ChannelType.PrivateThread"/>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:556]: Why is it always private?

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 17]
```diff
@@ -531,6 +551,28 @@ public Task ModifyEmbedSuppressionAsync(bool hideEmbeds)
...
+        /// </summary>
+        /// <param name="name">The name of the thread.</param>
+        /// <param name="auto_archive_duration"><see cref="ThreadAutoArchiveDuration"/> till it gets archived. Defaults to <see cref="ThreadAutoArchiveDuration.OneHour"/></param>
+        /// <param name="reason">The reason.</param>
+        /// <returns></returns>
+        /// <exception cref="Exceptions.UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.UsePrivateThreads"/> or <see cref="Permissions.SendMessages"/> permission.</exception>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:562]: client -> user

**Comment by Emzi0767** [comment]
In terms of scoping, I'd rather this was in 2 PRs for 2 reasons:
1. If this needs to be reverted, I'd rather deal with one change at once.
2. Because @DWaffles was already working on threads.

2 means that whenever waffles gets around to PRing his change, and we ended up going with his over yours (ultimately I'm gonna have to decide), it also reduces logistical problems.

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 18]
```diff
@@ -531,6 +551,28 @@ public Task ModifyEmbedSuppressionAsync(bool hideEmbeds)
...
+        /// <exception cref="Exceptions.NotFoundException">Thrown when the channel does not exist.</exception>
+        /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the <see cref="ThreadAutoArchiveDuration"/> cannot be modified or <see cref="ChannelType.PrivateThread"/> is not enabled for guild. This happens, if the guild does not have <see cref="PremiumTier.Tier_2"/>.</exception>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:566]: "Thrown when the guild does not have the required boost level for the specified <see cref="ThreadAutoArchiveDuration"/>" is more concise.

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 19]
```diff
@@ -660,6 +678,36 @@ public Task<DiscordChannel> CreateTextChannelAsync(string name, DiscordChannel p
...
+        /// <param name="reason">Reason for audit logs.</param>
+        /// <returns>The newly-created news channel.</returns>
+        /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.ManageChannels"/>.</exception>
+        /// <exception cref="NotFoundException">Thrown when the guild does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        /// <exception cref="NotSupportedException">Thrown when the guilds has not enabled community.</exception>
+        public Task<DiscordChannel> CreateNewsChannelAsync(string name, IEnumerable<DiscordOverwriteBuilder> overwrites = null, string reason = null)
+            => this.Features.Contains("COMMUNITY") ? this.CreateChannelAsync(name, ChannelType.News, null, Optional.FromNoValue<string>(), null, null, overwrites, null, Optional.FromNoValue<int?>(), null, reason) : throw new NotSupportedException("Guild has not enabled community. Can not create a news channel.");
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:709]: This should be InvalidOperationException or UnauthorizedException, I think.

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 20]
```diff
@@ -708,6 +756,16 @@ public Task<DiscordChannel> CreateChannelAsync(string name, ChannelType type, Di
                 : this.Discord.ApiClient.CreateGuildChannelAsync(this.Id, name, type, parent?.Id, topic, bitrate, userLimit, overwrites, nsfw, perUserRateLimit, qualityMode, reason);
         }
 
+        /// <summary>
+        /// Gets active threads. Can contain more threads. Watch for HasMore.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:760]: Read: Threads

File: DSharpPlus/Entities/Stage/DiscordStageInstance.cs [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+        /// Gets the topic of the Stage instance
+        /// </summary>
+        [JsonProperty("privacy_level", NullValueHandling = NullValueHandling.Ignore)]
+        public StagePrivacyLevel PrivacyLevel { get; internal set; }
+
+        /// <summary>
+        /// Gets whether or not stage discovery is disabled
+        /// </summary>
+        [JsonProperty("discoverable_disabled", NullValueHandling = NullValueHandling.Ignore)]
+        public bool DiscoverableDisabled { get; internal set; }
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Stage/DiscordStageInstance.cs:63]: Punctuation.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 22]
```diff
@@ -0,0 +1,690 @@
...
+                dto : null;
+
+        /// <summary>
+        /// Gets when the last pinned message was pinned in this thread as raw string.
+        /// </summary>
+        [JsonProperty("last_pin_timestamp", NullValueHandling = NullValueHandling.Ignore)]
+        internal string LastPinTimestampRaw { get; set; }
+
+        /// <summary>
+        /// Gets the threads metadata.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:121]: Read: redundant docs. "Gets metadata about the thread, e.g. [...]" something like that.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 23]
```diff
@@ -0,0 +1,690 @@
...
+using DSharpPlus;
+using DSharpPlus.Exceptions;
+using DSharpPlus.Net.Models;
+using DSharpPlus.Net.Serialization;
+using Newtonsoft.Json;
+
+namespace DSharpPlus.Entities
+{
+    /// <summary>
+    /// Represents a discord thread channel.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:40]: Worth noting that if the thread is started from a message, it's Id will be the same as the starter message.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 24]
```diff
@@ -0,0 +1,690 @@
...
+            => this.Type == ChannelType.PrivateThread;
+
+        /// <summary>
+        /// Gets the ID of the last message sent in this thread.
+        /// </summary>
+        [JsonProperty("last_message_id", NullValueHandling = NullValueHandling.Ignore)]
+        public ulong? LastMessageId { get; internal set; }
+
+        /// <summary>
+        /// <para>Gets the slow mode delay configured for this thread.</para>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:88]: slowmode is one word.

**Comment by Lulalaby** [comment]
That's quite unfair but well

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 25]
```diff
@@ -0,0 +1,690 @@
...
+        /// <para>All bots, as well as users with <see cref="Permissions.ManageChannels"/> or <see cref="Permissions.ManageMessages"/> permissions in the channel are exempt from slow mode.</para>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:89]: Users that can use this thread*

**Comment by Lulalaby** [comment]
> Because @DWaffles was already working on threads.

And this is not quite correct.
You know who started working on it and @DWaffles was working with _me_ on threads before #890 got closed because of the big amount of changes. Just to be clear about it.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 26]
```diff
@@ -0,0 +1,690 @@
...
+        /// Gets the members of a thread. Needs the <see cref="DiscordIntents.GuildMembers"/> intent.
+        /// </summary>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public async Task<IReadOnlyList<DiscordThreadChannelMember>> GetMembersAsync()
+            => await this.Discord.ApiClient.GetThreadMembersAsync(this.Id);
+
+        /// <summary>
+        /// Adds a thread member.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:253]: "Adds a member to this thread. Only applicable to private threads."

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 27]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <param name="member">The member to be added.</param>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task AddMemberAsync(DiscordMember member)
+            => this.AddMemberAsync(member.Id);
+
+        /// <summary>
+        /// Removes a thread member.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:273]: Removes a member from this thread.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 28]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <param name="member">The member id to be removed.</param>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task RemoveMemberAsync(ulong member)
+            => this.Discord.ApiClient.RemoveThreadMemberAsync(this.Id, member);
+
+        /// <summary>
+        /// Removes a thread member.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:283]: Ditto.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 29]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <param name="member">The member to be removed.</param>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task RemoveMemberAsync(DiscordMember member)
+            => this.RemoveMemberAsync(member.Id);
+
+        // ----------
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:292]: ??

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 30]
```diff
@@ -0,0 +1,690 @@
...
+        /// <summary>
+        /// Adds a role to a thread.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:294]: Adds all members with a given role to the current thread.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 31]
```diff
@@ -0,0 +1,690 @@
...
+            var members = await this.Guild.GetAllMembersAsync();
+            var roleMembers = members.Where(m => m.Roles.Contains(role));
+            foreach(var member in roleMembers)
+            {
+                await this.Discord.ApiClient.AddThreadMemberAsync(this.Id, member.Id);
+            }
+        }
+
+        /// <summary>
+        /// Adds a role to a thread.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:312]: Ditto.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 32]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <param name="role">The role to be added.</param>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task AddRoleAsync(DiscordRole role)
+            => this.AddRoleAsync(role.Id);
+
+        /// <summary>
+        /// Removes a role from a thread.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:322]: Ditto.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 33]
```diff
@@ -0,0 +1,690 @@
...
+            var members = await this.Guild.GetAllMembersAsync();
+            var roleMembers = members.Where(m => m.Roles.Contains(role));
+            foreach (var member in roleMembers)
+            {
+                await this.Discord.ApiClient.RemoveThreadMemberAsync(this.Id, member.Id);
+            }
+        }
+
+        /// <summary>
+        /// Removes a role from a thread.
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:340]: Ditto.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 34]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <param name="role">The role to be removed.</param>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task RemoveRoleAsync(DiscordRole role)
+            => this.RemoveRoleAsync(role.Id);
+
+        /// <summary>
+        /// Joins a thread
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:350]: Punctuation. Missing permissions remark.

**Comment by Lulalaby** [comment]
Well, you have the diff, you have my implementation now, do what you want.

I just wanted to be fair and submit my implementations to get this faster here.

**Comment by Emzi0767** [comment]
Let's not get into yet another fight. If you wish to coordinate thread contributions with waffles, feel free to; I'd be fine with this. 

As it stands, we have 2 implementations pending, of which one was done on my request. A choice based on merits sounds fair, it levels the playing field.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 35]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:353]: "Thrown when an invalid parameter was provided" is painfully vague.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 36]
```diff
@@ -0,0 +1,690 @@
...
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task JoinAsync()
+            => this.Discord.ApiClient.JoinThreadAsync(this.Id);
+
+        /// <summary>
+        /// Leaves a thread
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:359]: Punctuation.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 37]
```diff
@@ -0,0 +1,690 @@
...
+        public Task<DiscordMessage> SendMessageAsync(Action<DiscordMessageBuilder> action)
+        {
+            var builder = new DiscordMessageBuilder();
+            action(builder);
+
+            return this.Discord.ApiClient.CreateMessageAsync(this.Id, builder);
+        }
+
+        /// <summary>
+        /// Returns a specific message
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:448]: Punctuation.

File: DSharpPlus/Entities/Thread/DiscordThreadChannel.cs [hunk 38]
```diff
@@ -0,0 +1,690 @@
...
+        /// </summary>
+        /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.AccessChannels"/> or the <see cref="Permissions.ReadMessageHistory"/> permission.</exception>
+        /// <exception cref="NotFoundException">Thrown when the thread does not exist.</exception>
+        /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+        /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+        public Task<IReadOnlyList<DiscordMessage>> GetMessagesAsync(int limit = 100) =>
+            this.GetMessagesInternalAsync(limit, null, null, null);
+
+        /// <summary>
+        /// Returns a list of messages
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Thread/DiscordThreadChannel.cs:513]: Punctuation.

**Comment by Lulalaby** [comment]
@DWaffles feel free to take a look into  [dsp-master-lulalaby.diff.txt](https://github.com/DSharpPlus/DSharpPlus/files/6887735/dsp-master-lulalaby.diff.txt)
Maybe you miss something on your implementation and then you can take this as ref.
As now, it's fully working. Feel free to implement it into yours.

Have a nice week everyone

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]
Realized this got closed as I was reviewing. Nonetheless, I'd implore you spend just a smidge more time when writing documentation, among other things.

File: DSharpPlus/Entities/Channel/DiscordChannel.cs (hunk 9 above)
Comment (Lulalaby) [reply DSharpPlus/Entities/Channel/DiscordChannel.cs:648]: no this is discordchannel to create a public thread

File: DSharpPlus/Entities/Channel/DiscordChannel.cs (hunk 10 above)
Comment (Lulalaby) [reply DSharpPlus/Entities/Channel/DiscordChannel.cs:662]: watch for HasMore indicates that the api could have more and you have to recall the function

**Comment by Lulalaby** [comment]
Thanks for the feedback @VelvetThePanda 
I use it to fix this in my version.

File: DSharpPlus/Clients/DiscordClient.Dispatch.cs [hunk 39]
```diff
@@ -1604,6 +1661,197 @@ internal async Task OnMessageReactionRemoveEmojiAsync(ulong messageId, ulong cha
...
+        internal async Task OnThreadMembersUpdateEventAsync(DiscordGuild guild, ulong thread_id, IReadOnlyList<DiscordThreadChannelMember> addedMembers, IReadOnlyList<ulong?> removed_member_ids, int member_count)
+        {
+            guild.Discord = this;
+
+            var threadMembersUpdateArg = new ThreadMembersUpdateEventArgs
+            {
+                Guild = guild,
+                ThreadId = thread_id,
+                AddedMembers = addedMembers ?? Array.Empty<DiscordThreadChannelMember>(),
+                RemovedMemberIds = removed_member_ids ?? Array.Empty<ulong?>(),
```
[the remark this answers is not in the document]
Comment (Lulalaby) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1846]: What you mean by this?

[the remark this answers is not in the document]
Comment (Lulalaby) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1846]: RemovedMemberIds as DiscordMember?

[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1846]: Yes

[the remark this answers is not in the document]
Comment (Lulalaby) [reply DSharpPlus/Clients/DiscordClient.Dispatch.cs:1846]: ok I get you.

**Pull request closed without merging**